### ISSUE TITLE
computeParameters reads from memory that may contain garbage 

### ISSUE THREAD

**Issue by ex-rzr** [openmm/openmm#3392]
`TestOpenCLNonbondedForce` tests on AMD (and `TestHipNonbondedForce`) crash with segfaults, see https://github.com/openmm/openmm/pull/3386#issuecomment-997308486

This didn't happen before this change: 
https://github.com/openmm/openmm/blame/ed9df876d43c037c08d4762721e73e5caae086d9/platforms/opencl/tests/TestOpenCLNonbondedForce.cpp#L63-L66

With `addExceptionParameterOffset` `HAS_OFFSETS` is defined (but offsets are related to exclusions, not particles) and the kernel reads from `particleOffsetIndices` (and then `particleParamOffsets`):
https://github.com/openmm/openmm/blob/ed9df876d43c037c08d4762721e73e5caae086d9/platforms/common/src/kernels/nonbondedParameters.cc#L19-L21

This array is created but does not contain any meaningful value (`force.getNumParticleParameterOffsets() == 0`):
https://github.com/openmm/openmm/blob/ed9df876d43c037c08d4762721e73e5caae086d9/platforms/opencl/src/OpenCLKernels.cpp#L1057-L1060
https://github.com/openmm/openmm/blob/ed9df876d43c037c08d4762721e73e5caae086d9/platforms/cuda/src/CudaKernels.cpp#L1113-L1116

This test works with CUDA and OpenCL platforms on Nvidia but it doesn't mean that it have to work as both `cuMemAlloc` and `clCreateBuffer` do not **guarantee** that the allocated memory will be cleared: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g37d37965bfb4803b6d4e59ff26856356

In this test on AMD, these arrays contain garbage values from one of previous freed allocations (there are multiple `testParallelComputation` calls, that's why it works if only one of the sub-tests is called).

I managed to reproduce this crash on Nvidia with this garbage generator :)
```diff
diff --git a/platforms/cuda/src/CudaArray.cpp b/platforms/cuda/src/CudaArray.cpp
index 9d27e967d..ca8f27022 100644
--- a/platforms/cuda/src/CudaArray.cpp
+++ b/platforms/cuda/src/CudaArray.cpp
@@ -30,6 +30,7 @@
 #include <iostream>
 #include <sstream>
 #include <vector>
+#include <random>
 
 using namespace OpenMM;
 
@@ -67,6 +68,21 @@ void CudaArray::initialize(ComputeContext& context, int size, int elementSize, c
         str<<"Error creating array "<<name<<": "<<CudaContext::getErrorString(result)<<" ("<<result<<")";
         throw OpenMMException(str.str());
     }
+    size_t n = size*elementSize/sizeof(unsigned int);
+    if (n > 0) {
+        std::random_device rd;
+        std::mt19937 engine(rd());
+        std::vector<unsigned int> garbage(n);
+        for (size_t i = 0; i < n; i++) {
+            garbage[i] = engine();
+        }
+        result = cuMemcpyHtoD(pointer, garbage.data(), n*sizeof(unsigned int));
+        if (result != CUDA_SUCCESS) {
+            std::stringstream str;
+            str<<"Error creating array "<<name<<": "<<CudaContext::getErrorString(result)<<" ("<<result<<")";
+            throw OpenMMException(str.str());
+        }
+    }
 }
 
 void CudaArray::resize(int size) {
diff --git a/platforms/opencl/src/OpenCLArray.cpp b/platforms/opencl/src/OpenCLArray.cpp
index d9da1913e..916ebacbb 100644
--- a/platforms/opencl/src/OpenCLArray.cpp
+++ b/platforms/opencl/src/OpenCLArray.cpp
@@ -29,6 +29,7 @@
 #include <iostream>
 #include <sstream>
 #include <vector>
+#include <random>
 
 using namespace OpenMM;
 
@@ -69,6 +70,24 @@ void OpenCLArray::initialize(OpenCLContext& context, int size, int elementSize,
         str<<"Error creating array "<<name<<": "<<err.what()<<" ("<<err.err()<<")";
         throw OpenMMException(str.str());
     }
+    size_t n = size*elementSize/sizeof(unsigned int);
+    if (n > 0) {
+        std::random_device rd;
+        std::mt19937 engine(rd());
+        std::vector<unsigned int> garbage(n);
+        for (size_t i = 0; i < n; i++) {
+            garbage[i] = engine();
+        }
+        try {
+            context.getQueue().enqueueWriteBuffer(*buffer, CL_TRUE, 0, n*sizeof(unsigned int), garbage.data());
+            context.getQueue().finish();
+        }
+        catch (cl::Error err) {
+            std::stringstream str;
+            str<<"Error creating array "<<name<<": "<<err.what()<<" ("<<err.err()<<")";
+            throw OpenMMException(str.str());
+        }
+    }
 }
 
 void OpenCLArray::initialize(OpenCLContext& context, cl::Buffer* buffer, int size, int elementSize, const std::string& name) {
```

Luckily only these tests fail/crash with this change. So I assume this is the only kernel affected by the issue.

**Comment by peastman** [comment]
Thanks!  The fix is in #3399.  Could you try it out and see if it works correctly for you?

Sorry for taking a while to reply.  I'm just catching up after the holidays.

**Comment by ex-rzr** [comment]
@peastman 

I've just checked, it works. Thank you!
